### DISCUSSION TITLE
How can I create unique indexes?

### DISCUSSION THREAD

**Discussion by TenTakano** [TomGrozev/arangox_ecto#52] [Q&A] [answered]
Hi,

I'm trying to create an unique index for validating unique constraint.
I want to create a schema like the following and to add unique constraint on 'email' field. I know that unique index named '_key' is automatically created, but I don't want to use it for email field.

```
  schema "accounts" do
    field :email,           :string
    field :password,        :string, virtual: true, redact: true

    timestamps()
  end
```

I tried a migration file like the folloing, but unique index wasn't created. (I checked that by arangosh, and confirmed only primary key appeared.)

```
defmodule MyApp.Repo.Migrations.CreateAccountIndex do
  use ArangoXEcto.Migration

  def up do
    create(collection(:accounts))

    create(index(:accounts, [:email], unique: true))
  end
end
```

My environment is:

- elixir: 1.13.4-otp-24
- erlang: 24.3.4
- phoenix: 1.6.15
- ecto: 3.9.1
- ecto_sql: 3.9.0
- arangox: 0.5.4
- arangox_ecto: 1.2.0

**Comment by TomGrozev** [comment, marked as answer]
Hi, thanks for your question.

If you are using dynamic mode (default) I would recommend you check out https://hexdocs.pm/arangox_ecto/ArangoXEcto.Schema.html#indexes/1.

So you would have something like this:
```elixir

    indexes [
      [fields: [:email], unique: true]
    ]

    schema "accounts" do
      field :email,           :string
      field :password,        :string, virtual: true, redact: true

      timestamps()
    end

```

If you are using static mode then yes you will need migrations as you have done. Remember that migrations will only run up once, so if you have only added the index after the collection exit is it won't work, you will have to run `mix ecto.rollback` or similar.

A couple notes, firstly if you end up using `Ecto.Changeset.unique_constraint/3` make sure to provide a name using the `:name` option to both the index definition and the unique_constraint/3 function, otherwise you will get an error about an index name mismatch.

The second note is that you don't seem to be storing the password hash (unless you removed it in your post). If you haven't implemented it, here is one of many examples to get you started: https://stephanbv.medium.com/elixir-phoenix-lets-code-authentication-todo-application-part-1-599ee94cd04d#.vhr9cnt38
